### PR TITLE
1086/subscription error code

### PR DESCRIPTION
**Pull request by cornelinux**
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/privacyidea/privacyidea/pull/1087%23issuecomment-394065525%22%2C%20%22https%3A//github.com/privacyidea/privacyidea/pull/1087%23discussion_r192696797%22%2C%20%22https%3A//github.com/privacyidea/privacyidea/pull/1087%23discussion_r192715284%22%2C%20%22https%3A//github.com/privacyidea/privacyidea/pull/1087%23issuecomment-394384599%22%2C%20%22https%3A//github.com/privacyidea/privacyidea/pull/1087%23issuecomment-394469534%22%2C%20%22https%3A//github.com/privacyidea/privacyidea/pull/1087%23issuecomment-394611081%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/privacyidea/privacyidea/pull/1087%23issuecomment-394065525%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22%23%20%5BCodecov%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1087%3Fsrc%3Dpr%26el%3Dh1%29%20Report%5Cn%3E%20Merging%20%5B%231087%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1087%3Fsrc%3Dpr%26el%3Ddesc%29%20into%20%5Bmaster%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/commit/625a1213b98c4306eac2b7d4017891edc1c8b7ec%3Fsrc%3Dpr%26el%3Ddesc%29%20will%20%2A%2Aincrease%2A%2A%20coverage%20by%20%60%3C.01%25%60.%5Cn%3E%20The%20diff%20coverage%20is%20%60100%25%60.%5Cn%5Cn%5B%21%5BImpacted%20file%20tree%20graph%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1087/graphs/tree.svg%3Fwidth%3D650%26height%3D150%26src%3Dpr%26token%3D7nHLZki60B%29%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1087%3Fsrc%3Dpr%26el%3Dtree%29%5Cn%5Cn%60%60%60diff%5Cn%40%40%20%20%20%20%20%20%20%20%20%20%20%20Coverage%20Diff%20%20%20%20%20%20%20%20%20%20%20%20%20%40%40%5Cn%23%23%20%20%20%20%20%20%20%20%20%20%20master%20%20%20%20%231087%20%20%20%20%20%20%2B/-%20%20%20%23%23%5Cn%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%5Cn%2B%20Coverage%20%20%2095.62%25%20%20%2095.63%25%20%20%20%2B%3C.01%25%20%20%20%20%20%5Cn%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%5Cn%20%20Files%20%20%20%20%20%20%20%20%20131%20%20%20%20%20%20131%20%20%20%20%20%20%20%20%20%20%20%20%20%20%5Cn%20%20Lines%20%20%20%20%20%20%2016280%20%20%20%2016293%20%20%20%20%20%20%2B13%20%20%20%20%20%5Cn%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%3D%5Cn%2B%20Hits%20%20%20%20%20%20%20%2015568%20%20%20%2015581%20%20%20%20%20%20%2B13%20%20%20%20%20%5Cn%20%20Misses%20%20%20%20%20%20%20%20712%20%20%20%20%20%20712%5Cn%60%60%60%5Cn%5Cn%5Cn%7C%20%5BImpacted%20Files%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1087%3Fsrc%3Dpr%26el%3Dtree%29%20%7C%20Coverage%20%5Cu0394%20%7C%20%7C%5Cn%7C---%7C---%7C---%7C%5Cn%7C%20%5Bprivacyidea/lib/error.py%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1087/diff%3Fsrc%3Dpr%26el%3Dtree%23diff-cHJpdmFjeWlkZWEvbGliL2Vycm9yLnB5%29%20%7C%20%6088.46%25%20%3C100%25%3E%20%28%2B1.64%25%29%60%20%7C%20%3Aarrow_up%3A%20%7C%5Cn%7C%20%5Bprivacyidea/api/before%5C%5C_after.py%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1087/diff%3Fsrc%3Dpr%26el%3Dtree%23diff-cHJpdmFjeWlkZWEvYXBpL2JlZm9yZV9hZnRlci5weQ%3D%3D%29%20%7C%20%6097.34%25%20%3C100%25%3E%20%28%5Cu00f8%29%60%20%7C%20%3Aarrow_up%3A%20%7C%5Cn%5Cn------%5Cn%5Cn%5BContinue%20to%20review%20full%20report%20at%20Codecov%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1087%3Fsrc%3Dpr%26el%3Dcontinue%29.%5Cn%3E%20%2A%2ALegend%2A%2A%20-%20%5BClick%20here%20to%20learn%20more%5D%28https%3A//docs.codecov.io/docs/codecov-delta%29%5Cn%3E%20%60%5Cu0394%20%3D%20absolute%20%3Crelative%3E%20%28impact%29%60%2C%20%60%5Cu00f8%20%3D%20not%20affected%60%2C%20%60%3F%20%3D%20missing%20data%60%5Cn%3E%20Powered%20by%20%5BCodecov%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1087%3Fsrc%3Dpr%26el%3Dfooter%29.%20Last%20update%20%5B625a121...a4895d6%5D%28https%3A//codecov.io/gh/privacyidea/privacyidea/pull/1087%3Fsrc%3Dpr%26el%3Dlastupdated%29.%20Read%20the%20%5Bcomment%20docs%5D%28https%3A//docs.codecov.io/docs/pull-request-comments%29.%5Cn%22%2C%20%22created_at%22%3A%20%222018-06-02T07%3A17%3A10Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/in/254%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/apps/codecov%22%7D%7D%2C%20%7B%22body%22%3A%20%22Overall%2C%20I%20like%20this%20PR%20because%20it%20makes%20error%20handling%20more%20consistent%21%5Cr%5CnHowever%2C%20I%20saw%20that%20there%20are%20still%20some%20instances%20in%20which%20negative%20error%20codes%20are%20passed%20to%20%60%60send_error%60%60%2C%20e.g.%20in%20%60%60auth_error%60%60%3A%20https%3A//github.com/privacyidea/privacyidea/blob/a4895d6a2ba26fb39d59cda28b18cbffc7b30dc8/privacyidea/api/before_after.py%23L237%5Cr%5Cn%5Cr%5Cn...%20do%20we%20also%20want%20to%20change%20this%20error%20code%20here%3F%20Or%20should%20these%20stay%20negative%3F%22%2C%20%22created_at%22%3A%20%222018-06-04T14%3A56%3A10Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/28243%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/fredreichbier%22%7D%7D%2C%20%7B%22body%22%3A%20%22I%20think%20you%20are%20right.%20The%20thing%20is%2C%20we%20have%20a%20%60%60BaseError%60%60%20and%20a%20%60%60privacyIDEAError%60%60.%20I%20am%20wondering%20why%20we%20have%20a%20%60%60BaseError%60%60%20which%20is%20the%20base%20class%20only%20for%20the%20%60%60AuthError%60%60.%5Cr%5Cn%5Cr%5CnWe%20could%20change%20this%20all%20to%20remove%20the%20%60%60BaseError%60%60%20and%20also%20inherit%20the%20%60%60AuthError%60%60%20from%20the%20%60%60privacyIDEAError%60%60.%22%2C%20%22created_at%22%3A%20%222018-06-04T19%3A25%3A31Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars1.githubusercontent.com/u/1908620%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cornelinux%22%7D%7D%2C%20%7B%22body%22%3A%20%22I%20have%20wondered%20about%20this%20as%20well.%20I%20think%20one%20difference%20is%20that%20one%20can%20set%20the%20HTTP%20status%20code%20of%20an%20%60%60AuthError%60%60.%20However%2C%20it%20seems%20like%20all%20current%20%60%60AuthError%60%60%20instances%20have%20error%20code%20401%20anway.%22%2C%20%22created_at%22%3A%20%222018-06-05T07%3A33%3A47Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/28243%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/fredreichbier%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%20a4895d6a2ba26fb39d59cda28b18cbffc7b30dc8%20privacyidea/lib/error.py%2016%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/privacyidea/privacyidea/pull/1087%23discussion_r192696797%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22One%20question%3A%20Do%20I%20understand%20correctly%20that%20these%20do%20not%20correspond%20to%20HTTP%20error%20codes%2C%20these%20are%20just%20the%20codes%20used%20in%20the%20privacyIDEA%20response%3F%5Cr%5Cn%22%2C%20%22created_at%22%3A%20%222018-06-04T10%3A36%3A51Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars2.githubusercontent.com/u/28243%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/fredreichbier%22%7D%7D%2C%20%7B%22body%22%3A%20%22Yes.%20These%20are%20error%20codes%20in%20the%20JSON%20response%5Cr%5Cn%5Cr%5Cn%20%20%20%20%7B%20status%3A%20False%2C%5Cr%5Cn%20%20%20%20%20%20error%3A%20%7B%20code%3A%20...%2C%5Cr%5Cn%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20message%3A%20....%20%7D%5Cr%5Cn%5Cr%5Cn%22%2C%20%22created_at%22%3A%20%222018-06-04T11%3A56%3A22Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars1.githubusercontent.com/u/1908620%3Fv%3D4%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/cornelinux%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20privacyidea/lib/error.py%3AL33-54%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/privacyidea/privacyidea/pull/1087#issuecomment-394065525'>General Comment</a></b>
- <a href='https://github.com/apps/codecov'></a> # [Codecov](https://codecov.io/gh/privacyidea/privacyidea/pull/1087?src=pr&el=h1) Report
> Merging [#1087](https://codecov.io/gh/privacyidea/privacyidea/pull/1087?src=pr&el=desc) into [master](https://codecov.io/gh/privacyidea/privacyidea/commit/625a1213b98c4306eac2b7d4017891edc1c8b7ec?src=pr&el=desc) will **increase** coverage by `<.01%`.
> The diff coverage is `100%`.
[Impacted file tree graph](https://codecov.io/gh/privacyidea/privacyidea/pull/1087?src=pr&el=tree)
```diff
@@            Coverage Diff             @@
##           master    #1087      +/-   ##
==========================================
+ Coverage   95.62%   95.63%   +<.01%
==========================================
Files         131      131
Lines       16280    16293      +13
==========================================
+ Hits        15568    15581      +13
Misses        712      712
```
| [Impacted Files](https://codecov.io/gh/privacyidea/privacyidea/pull/1087?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [privacyidea/lib/error.py](https://codecov.io/gh/privacyidea/privacyidea/pull/1087/diff?src=pr&el=tree#diff-cHJpdmFjeWlkZWEvbGliL2Vycm9yLnB5) | `88.46% <100%> (+1.64%)` | :arrow_up: |
| [privacyidea/api/before\_after.py](https://codecov.io/gh/privacyidea/privacyidea/pull/1087/diff?src=pr&el=tree#diff-cHJpdmFjeWlkZWEvYXBpL2JlZm9yZV9hZnRlci5weQ==) | `97.34% <100%> (ø)` | :arrow_up: |
------
[Continue to review full report at Codecov](https://codecov.io/gh/privacyidea/privacyidea/pull/1087?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/privacyidea/privacyidea/pull/1087?src=pr&el=footer). Last update [625a121...a4895d6](https://codecov.io/gh/privacyidea/privacyidea/pull/1087?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).
- <a href='https://github.com/fredreichbier'></a> Overall, I like this PR because it makes error handling more consistent!
However, I saw that there are still some instances in which negative error codes are passed to ``send_error``, e.g. in ``auth_error``: https://github.com/privacyidea/privacyidea/blob/a4895d6a2ba26fb39d59cda28b18cbffc7b30dc8/privacyidea/api/before_after.py#L237
... do we also want to change this error code here? Or should these stay negative?
- <a href='https://github.com/cornelinux'></a> I think you are right. The thing is, we have a ``BaseError`` and a ``privacyIDEAError``. I am wondering why we have a ``BaseError`` which is the base class only for the ``AuthError``.
We could change this all to remove the ``BaseError`` and also inherit the ``AuthError`` from the ``privacyIDEAError``.
- <a href='https://github.com/fredreichbier'></a> I have wondered about this as well. I think one difference is that one can set the HTTP status code of an ``AuthError``. However, it seems like all current ``AuthError`` instances have error code 401 anway.
- [ ] <a href='#crh-comment-Pull a4895d6a2ba26fb39d59cda28b18cbffc7b30dc8 privacyidea/lib/error.py 16'></a> &nbsp;<b><a href='https://github.com/privacyidea/privacyidea/pull/1087#discussion_r192696797'>File: privacyidea/lib/error.py:L33-54</a></b>
- <a href='https://github.com/fredreichbier'></a> One question: Do I understand correctly that these do not correspond to HTTP error codes, these are just the codes used in the privacyIDEA response?
- <a href='https://github.com/cornelinux'></a> Yes. These are error codes in the JSON response
{ status: False,
error: { code: ...,
message: .... }

<a href='https://www.codereviewhub.com/privacyidea/privacyidea/pull/1087?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/privacyidea/privacyidea/pull/1087?approve=1'></a>&nbsp;<a href='https://github.com/privacyidea/privacyidea/pull/1087'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by cornelinux** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/privacyidea/privacyidea/pull/1087'>Refresh</a>.<a href='#crh-update-none'></a>

File: privacyidea/lib/error.py [hunk 1]
```diff
@@ -33,6 +33,21 @@
...
+    CONFIGADMIN = 302
+    POLICY = 303
+    VALIDATE = 401
+    REGISTRATION = 402
+    CA = 503
+    HSM = 707
+    SELFSERVICE = 807
+    SERVER = 903
+    USER = 904
+    PARAMETER = 905
```
Comment (fredreichbier) [inline privacyidea/lib/error.py]: One question: Do I understand correctly that these do not correspond to HTTP error codes, these are just the codes used in the privacyIDEA response?

Comment (cornelinux) [reply privacyidea/lib/error.py]: Yes. These are error codes in the JSON response

    { status: False,
      error: { code: ...,
                  message: .... }

**Comment by fredreichbier** [comment]
Overall, I like this PR because it makes error handling more consistent!
However, I saw that there are still some instances in which negative error codes are passed to ``send_error``, e.g. in ``auth_error``: https://github.com/privacyidea/privacyidea/blob/a4895d6a2ba26fb39d59cda28b18cbffc7b30dc8/privacyidea/api/before_after.py#L237

... do we also want to change this error code here? Or should these stay negative?

**Comment by fredreichbier** [comment]
I have wondered about this as well. I think one difference is that one can set the HTTP status code of an ``AuthError``. However, it seems like all current ``AuthError`` instances have error code 401 anway.

**Pull request merged**